### PR TITLE
Allow only one subscription per affordance

### PR DESCRIPTION
**Pull request by relu91**
This PR adds an additional step to `observeProperty` and `subscribeEvent` algorithms of the `ConsumeThing`. It forces the runtime implementation to check whether the affordance has already an active Subscription, and if it is the case the runtime must throw a `NotAllowedError`. This should make clear to implementers that it is not possible to subscribe multime times to the same affordance, but I also added an editor note to make it cristal clear. 

The PR is currently in draft mode cause I wanted to use [Set](https://tc39.es/ecma262/#sec-set-objects) as type of the list of active Subscriptions/Observations but ECMA specification is not included among our `xref` and it conflicts with `web-platform`. 

*Note*: It covers the points raised in #346.

***
<a href="https://pr-preview.s3.amazonaws.com/relu91/wot-scripting-api/pull/356.html" title="Last updated on Dec 13, 2021, 1:08 PM UTC (667e9c9)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/wot-scripting-api/356/ec3b3fc...relu91:667e9c9.html" title="Last updated on Dec 13, 2021, 1:08 PM UTC (667e9c9)">Diff</a>

### CODE REVIEW

File: index.html [hunk 1]
```diff
@@ -1213,6 +1213,16 @@ <h4>Internal slots for ConsumedThing</h4>
           <td>`null`</td>
           <td>The <a>Thing Description</a> of the {{ConsumedThing}}.</td>
          </tr>
+        <tr>
+          <td><dfn>[[\activeSubscriptions]]</dfn></td>
+          <td>`{}`</td>
```
Comment (zolkis) [inline index.html:1218]: It was a good idea to have an internal slot as a set, as it is usable with all policies (replace existing, reject if existing, add, ...).

File: index.html [hunk 2]
```diff
@@ -1739,8 +1764,16 @@ <h3>The <dfn>subscribeEvent()</dfn> method</h3>
         Takes as arguments |eventName:string|, |listener:WoTListener| and
         optionally  |onerror:ErrorListener| and |options:InteractionOptions|.
         It returns a {{Promise}} to signal success or failure.
+        <p class="ednote">
+          This algorithm allow for only one active {{Subscription}} per <a>Event</a>. If a new
```
Comment (zolkis) [inline index.html:1768]: Should we repeat this every time, or just say once for the whole API?

**Review by zolkis — APPROVED** [review]
Good work, thanks!

File: index.html (hunk 2 above)
Comment (relu91) [reply index.html:1768]: Good point. I think another good place to state this is in the `Subscription` section but I don't know if we want to keep this simple to possibly change it in the future.

**Review by danielpeintner — COMMENTED** [review]
I am wondering why this PR includes changes to 
 typescript/thing-description/schema/td-json-schema-validation.json
and 
 typescript/thing-description/thing-description.d.ts 

The changes should be already there... sync with master ?

File: index.html (hunk 2 above)
Comment (zolkis) [reply index.html:1768]: We can also omit the ednote, since the algorithm is clear enough.

**Comment by relu91** [comment]
Cleaned up commit history and added a fix for #358.

File: index.html [hunk 3]
```diff
@@ -1310,6 +1310,16 @@ <h4>Internal slots for ConsumedThing</h4>
...
          </tr>
+        <tr>
+          <td><dfn>[[\activeSubscriptions]]</dfn></td>
+          <td>`{}`</td>
+          <td>A {{Set}} containing the names of active {{Subscription}}'s for an <a>Event</a></td>
+        </tr>
+        <tr>
+          <td><dfn>[[\activeObservations]]</dfn></td>
+          <td>`{}`</td>
+          <td>A {{Set}} containing the names of active {{Subscription}}'s for an <a>Property</a></td>
```
Comment (danielpeintner) [inline index.html:1321]: I guess it should read 
a set of active Observations but then internal linking is no longer  working.... maybe we can refer to it by id

File: index.html [hunk 4]
```diff
@@ -1677,8 +1687,16 @@ <h3>The <dfn>observeProperty()</dfn> method</h3>
         Takes as arguments |propertyName:string|, |listener:InteractionListener| and
         optionally |onerror:ErrorListener| and |options:InteractionOptions|.
         It returns a {{Promise}} that resolves on success and rejects on failure.
+        <p class="ednote">
+          This algorithm allow for only one active {{Subscription}} per <a>Property</a>. If a new
```
Comment (danielpeintner) [inline index.html:1691]: typo?: this algorithm allowS ...

File: index.html (hunk 2 above)
Comment (danielpeintner) [reply index.html:1865]: typo?: this algorithm allowS ...

**Comment by danielpeintner** [comment]
Scripting Call 2021-12-13
@relu91 fixes the 2 editorial issues and than we can merge

**Comment by relu91** [comment]
@danielpeintner it should be ready to go :)

**Review by danielpeintner — APPROVED** [review]

**Pull request merged**